### PR TITLE
The Enumerable Test Cases are Created

### PR DESCRIPTION
**Pull request by SafaErden**
We have created test cases for all enumerable methods and now the spec branch is ready to be reviewed.

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  let(:test1) { [5, 4, 2] }
+  let(:test2) { [3] }
+  let(:test3) { [nil, true, 10] }
+  let(:test4) { [1, 2, 3, 1, 2, 3, 4, 1, 2, 3] }
+
+  describe '#my_each' do
+    it 'should interate through an array' do
+      expect(test1.my_each { |x| p x }).to eql(test1.each { |x| p x })
+      expect(test1.my_each { |x| p x * 2 }).to eql(test1.each { |x| p x * 2 })
+      expect(test1.my_each { |x| p x * 4 }).to eql(test1.each { |x| p x * 4 })
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe '#my_each_with_index' do
+    it 'should interate through an array and its index' do
+      expect(test1.my_each_with_index { |x, y| x + y }).to eql(test1.each_with_index { |x, y| x + y })
+      expect(test1.my_each_with_index { |x, y| x * y }).to eql(test1.each_with_index { |x, y| x * y })
+      expect(test1.my_each_with_index { |x, y| x - y }).to eql(test1.each_with_index { |x, y| x - y })
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe '#my_select' do
+    it 'should create a new array based on the conditions passed at the block' do
+      expect(test2.my_select { |x| x == 1 }).to eql([])
+      expect(test3.my_select { |x| x == true }).to eql([true])
+      expect(test4.my_select { |x| x == 1 }).to eql([1, 1, 1])
+      expect(test4.my_select { |x| x > 3 }).to eql([4])
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe '#my_all?' do
+    it 'should check when the given condition is true for the itens in the array' do
+      expect(test1.my_all? { |x| x == 1 }).to eql(false)
+      expect(test1.my_all? { |x| x > 0 }).to eql(true)
+      expect(test1.my_all? { |x| x == true }).to eql(false)
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe '#my_any?' do
+    it 'should check whether the given condition is true for every item in the array' do
+      expect(test3.my_any? { |x| x == true }).to eql(true)
+      expect(test4.my_any? { |x| x == true }).to eql(false)
+      expect(test4.my_any? { |x| x > 0 }).to eql(true)
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  describe '#my_none?' do
+    it 'should check if the given condition is true for every item in the array' do
+      expect(test3.my_none? { |x| x == true }).to eql(false)
+      expect(test4.my_none? { |x| x == true }).to eql(true)
+      expect(test4.my_none? { |x| x > 0 }).to eql(false)
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: You should have only one `it` block per `expected` behavior ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png)  from  Pathwright).

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+require_relative '../my_methods.rb'
+
+RSpec.describe Enumerable do
+  let(:test1) { [5, 4, 2] }
+  let(:test2) { [3] }
+  let(:test3) { [nil, true, 10] }
+  let(:test4) { [1, 2, 3, 1, 2, 3, 4, 1, 2, 3] }
+
+  describe '#my_each' do
+    it 'should interate through an array' do
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: The correct word is "iterate". Please fix this on all of your tests.

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
+# spec/enumerables_spec.rb
+
+require_relative '../my_methods.rb'
+
+RSpec.describe Enumerable do
+  let(:test1) { [5, 4, 2] }
+  let(:test2) { [3] }
+  let(:test3) { [nil, true, 10] }
+  let(:test4) { [1, 2, 3, 1, 2, 3, 4, 1, 2, 3] }
```
Comment (majovanilla) [inline spec/enumerable_spec.rb:9]: Please include other types of enumerable in your testing variables, like hashes or ranges.

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      end
+      it 'Array.new(10) { rand(0...10).my_inject(:+)' do
+        expect(array.my_inject(:+)).to eql(array.inject(:+))
+      end
+      it 'Array.new(10) { rand(0...10).my_inject(2,:*)' do
+        expect(array.my_inject(2, :*)).to eql(array.inject(2, :*))
+      end
+      it 'Array.new(10) { rand(0...10).my_inject(2){|product, n| product * n}' do
+        expect(array.my_inject(2) { |product, n| product * n }).to eql(array.inject(2) { |product, n| product * n })
+      end
```
Comment (majovanilla) [inline spec/enumerable_spec.rb]: The name of the `it` block it's not descriptive of what it's testing. Please improve the readability of your tests.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hi team!

## You did a *good* job on this project! :+1: 

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Update your Readme file. It should include the testing instructions.

- You have some extra code on your my_methods.rb that is causing Rspec to show a strange output. Please delete that to avoid confusion.

- You should test you methods to see if they return an enumerator when they have to.

- Right now, you are testing for the "perfect" behavior hoping nothing goes wrong. You should also consider "the worst-case scenario" and try for negative test cases to check that nothing unexpected happens. For example, if you pass the wrong number of arguments to a method.

- Check the commented code below and implement the changes.

### Optional Changes:

If you want to make your project even better 🔥💥, I have a few suggestions you can implement.  

These are completely [Optional]: 🦔

- I noticed that you use `each` on your `my_each` method. This is not ok since the purpose of the project was to develop the methods on your own. You could change this for a loop of your choice and then test your methods to see if they pass the corresponding tests. If you do that, you would have to change `my_each_with_index` too, because it will no longer work.

- You should add `--format documentation` to your `.rspec` file to have a detailed output of all the tests.

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request Under the review by Maria Eugenia Queme P. (TSE)

### Hello @SafaErden and @kevinfrontend :balloon: :lollipop: 

### Thank you for doing the changes requested by @majovanilla, your project looks more professional now :medal_sports: :medal_sports: and overall this is good progress :muscle: there are still some changes to work on :construction: :construction_worker: :construction_worker: 

#### General

- [x] As it was requested by the previous reviewer you haven't tested other Enumerable like Ranges or Hashes let's add them

- [x] I see your Description for the `it` blocks within my_all, my_any and my_none is the same, be more specific, for each of the methods, they do different things for example:

your test case now:
```ruby
  describe '#my_all?' do
    it 'should check when the given condition is true for the itens in the array' do
      expect(test1.my_all? { |x| x == 1 }).to eql(test1.all? { |x| x == 1 })
    end
  end
```
your test case with a more descriptive `it` block
```ruby
describe '#my_all?' do
    it 'should return true if all of the elements of the Enumerable pass the condition given by the block' do
      expect(test1.my_all? { |x| x == 1 }).to eql(true)
    end
  end
```
#### Project-specific 

- [x] specs cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- I see you still haven't added negative test cases, each of your Enumerable describe blocks should contain and `it` block with a negative test case, but for the purpose of the project let's add them at least for the methods that retrieve a boolean (my_all, my_any and my_none)

here's an example:

Your code originally
```ruby
  describe '#my_any?' do
    it 'should check whether the given condition is true for every item in the array' do
      expect(test3.my_any? { |x| x == true }).to eql(true)
    end
  end
```
Adding a negative test case (with a more descriptive it block for the positive test case)

```ruby
let(:test5) { [false, false, false] }
  describe '#my_any?' do
    it 'should return true if at least one of the elements of the Enumerable pass the condition given by the block' do
      expect(test3.my_any? { |x| x == true }).to eql(true)
    end
   it "shouldn't return true if none of the elements of the Enumerable pass the condition given by the block" do
      expect(test5.my_any? { |x| x == true }).not_to eql(true)
    end
  end
```
- As you can see the it block starts with a negative word and I've used the [not_to function](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers)

[OPTIONAL]
- [ ] All methods implemented in the Enumerable project are tested ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/all_enumerables.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- I'm sorry this wasn't pointed out in the previous review but this is a hard requirement, all Enumerable methods have to be tested, you are missing the test case for my_count, I'm setting this as optional but I strongly advise working on it

### That's it, team :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? ask me on slack **@ Maru**

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Comment by kevcap** [comment]
Thanks for your considerations! We'll work on it.

**Review by mikenath223 — APPROVED** [review]
### CONGRATULATIONS  👏👏👏

This project meets the requirements for approval.

**STATUS: APPROVED ✅**

- [ ] OPTIONAL - Consider the fact that my_any is the opposite of my_none, it's good practice to use a shorter method to achieve a goal.

Don't forget to fill the project completion form👍
Good luck with the next project.

HAPPY CODING👨‍💻

**Pull request merged**